project-koku/koku
# [COST-3484] Process ROS Report

## Jira Ticket

[COST-3484](https://issues.redhat.com/browse/COST-3484)

## Description

This change will collect ROS reports during payload processing, send them to S3 for consumption by the ROS team and send a kafka message containing relevant information and the processed files. 

## Testing
Testing for this really has to be done in ephemeral, please reach out if you need help through this part.
1. Checkout Branch
2. Reserve an ephemeral namespace
3. Deploy koku to ephemeral
4. After it has deployed, create a source and send a payload to ingress for an operator report that contains ROS metrics. (if you need help with this part, reach out and I can help)
5. Monitor the processing flow and make sure everything processes appropriately. Check minio and the kafka messages in ephemeral to ensure they are correct.


Sample kafka message, the URL is generated by [an s3 client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/generate_presigned_url.html#generate-presigned-url).
```
{
  "request_id": "ingress-service-665ff59b4b-jxx2z/naZpbyUQCy-001151",
  "b64_identity": "eyJpZGVudGl0eSI6IHsib3JnX2lkIjogIjEyMzQ1NjciLCAiYWNjb3VudF9udW1iZXIiOiAiMTAwMDEiLCAidHlwZSI6ICJVc2VyIiwgInVzZXIiOiB7InVzZXJuYW1lIjogInVzZXJfZGV2IiwgImVtYWlsIjogInVzZXJfZGV2QGZvby5jb20iLCAiaXNfb3JnX2FkbWluIjogdHJ1ZSwgImFjY2VzcyI6IHt9fX0sImVudGl0bGVtZW50cyI6IHsiY29zdF9tYW5hZ2VtZW50IjogeyJpc19lbnRpdGxlZCI6IHRydWV9fX0=",
  "metadata": {
    "account": "10001",
    "org_id": "1234567",
    "source_id": "c40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1",
    "cluster_uuid": "023d9b0e-7ca6-481d-b04f-ea606becd54e",
    "cluster_alias": "ocp-ros-testing"
  },
  "files": [
    "http://env-ephemeral-y4ldf0-minio.ephemeral-y4ldf0.svc:9000/ros-eph-s3/org1234567/source%3Dc40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date%3D2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.4.csv?AWSAccessKeyId=<key>&Signature=<signature>&Expires=1680267516",
    "http://env-ephemeral-y4ldf0-minio.ephemeral-y4ldf0.svc:9000/ros-eph-s3/org1234567/source%3Dc40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date%3D2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.5.csv?AWSAccessKeyId=<key>&Signature=<signature>&Expires=1680267516"
  ],
  "object_keys": [
    "org1234567/source=c40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date=2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.4.csv",
    "org1234567/source=c40f7c0a-c96b-4d33-aebe-5a1ccc1c5ad1/date=2023-03-29/008f31ac-67e2-4682-bf87-69ba400584e1_openshift_usage_report.5.csv"
  ]
}
```
Example LOGs from Ephemeral:
When Processing just koku reports:
```
[2023-03-27 12:55:03,561] INFO None {'message': 'No ROS reports to handle for manifest: 1', 'tracing_id': 'ingress-service-6bd86c44d5-8ntq6/eVSWrsjHG2-000407', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': 'b97fe3a1-26bd-4489-b8fb-8146a4187897', 'provider_type': 'OCP'}
```

When Processing ROS reports:
```
[2023-03-27 12:56:22,085] INFO None {'message': 'Preparing to upload ROS reports to S3 bucket for manifest: 2.', 'tracing_id': 'ingress-service-6bd86c44d5-8ntq6/eVSWrsjHG2-000438', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': 'b97fe3a1-26bd-4489-b8fb-8146a4187897', 'provider_type': 'OCP'}
[2023-03-27 12:56:22,432] INFO None {'message': '1 reports uploaded to S3 for ROS, sending kafka message.', 'tracing_id': 'ingress-service-6bd86c44d5-8ntq6/eVSWrsjHG2-000438', 'account': '10001', 'org_id': '1234567', 'cluster_id': '023d9b0e-7ca6-481d-b04f-ea606becd54e', 'provider_uuid': 'b97fe3a1-26bd-4489-b8fb-8146a4187897', 'provider_type': 'OCP'}
[2023-03-27 12:56:22,437] INFO None kafka message delivered.
```
## Notes

...
